### ISSUE TITLE
improve the application registry for use in packages

### ISSUE THREAD

**Issue by keewis** [hgrecco/pint#1365]
xref xarray-contrib/pint-xarray#129

The currently recommended way to use the application registry is:
```python
# application
ureg = pint.UnitRegistry()
# configuration
pint.set_application_registry(ureg)

import package_using_pint

# package_using_pint
registry = pint.get_application_registry()
```

There are a few issues with this:
1. there's code before an import so tools like `isort` will complain
2. since all `set_application_registry` does is assigning its argument to a module-global, all other references to the previous object are not updated.
3. as a consequence of 2, packages providing new registries either need to be imported before packages using them, or they need to provide a setup function that needs to be called before the import of those packages

I don't really like setup functions, though, especially if the package can clearly separate the unit registry from other functionality. E.g. `import cf_xarray` does not even try to import `pint`, but `import cf_xarray.units` will create and set a custom unit registry.

To solve this, I'd like to propose exposing a `UnitRegistry` wrapper object as `pint.application_registry`:
```python
class ApplicationRegistry:
    def __init__(self, registry):
        self._registry = registry

    def set(self, new):
        self._registry = new

    def __getattr__(self, name):
        return getattr(self._registry, name)

    def __dir__(self):
        return dir(self._registry)

    # other __*__ methods

application_registry = ApplicationRegistry(_DEFAULT_REGISTRY)
```
with this, packages using the application registry can just `import pint` or `from pint import application_registry`, and other packages providing a registry can call `.set` to have all other packages get the new registry, without having to care when/in which order they are imported.

Also note that this is completely independent of the `get/set_application_registry` functions.

Not sure if I'm missing something obvious, but if this sounds like a good idea I'll send in a PR.

**Comment by jthielen** [comment]
For what it's worth, I like this idea! [MetPy can be added to the list of downstream packages where something like this would be useful](https://github.com/Unidata/MetPy/issues/1350). Likewise, I'm not sure if I'm missing something either, but I did have one concern: the naming of it. It might be too confusing to have two independent things both referred to as an application registry. If they really are independent, then maybe this should be called something different (perhaps `SharedRegistry`)? Or, if this proves to be the better approach, perhaps `get/set_application_registry` can be deprecated and defer to this `ApplicationRegistry` for now?

**Comment by keewis** [comment]
> It might be too confusing to have two independent things both referred to as an application registry.

What I meant with "independent" was that the proposal on its own would be fully backwards compatible. I agree that having two different things have the same name is confusing, which is why I would probably deprecate `get/set_application_registry` or rewrite them to use the `ApplicationRegistry` object internally (not sure if a deprecation cycle would be needed for that).

**Comment by hgrecco** [comment]
In general I like the approach as it make it more robust to share a registry across packages using pint.  As the pint ecosysem is growing, this is extemely important now more than ever. So I am +1 on rethinking the current approach.

Something I liked about the `get/set_application_registry` is that if a package did not care about having a registry of their own, instead of doing:
```python
ureg = pint.UnitRegistry()
```
they just do:
```python
ureg = pint.get_application_registry()
```
which will create a registry if needed (but return the application registry if it has been set). This is something that I would like to keep.

But I do not like this
```python
ureg = pint.application_registry
```
because it makes it very easy to change (and break `pint.application_registry` variable)

I would rather rewrite `get/set_application_registry` using your `ApplicationRegistry`. What do you think?

**Comment by keewis** [comment]
I was thinking that `from pint import application_registry` would somehow be the preferred way to do this, but we can of course rewrite `get/set_application_registry` to use `ApplicationRegistry` internally.

I'll create a PR, then, and we can continue the discussion there.

**Comment by dopplershift** [comment]
@keewis Was the omission of `__setattr__` here intentional?

**Comment by keewis** [comment]
no, of course not. The fix has already been merged (#1403) and we're just waiting on a release.

**Comment by jules-ch** [comment]
@hgrecco can we make a release? At least for the bug fixed in #1403
